[GSoC][JAX FE] Initialize the prototype for the jax frontend

AsakusaRinne opened the pull request:
This PR is going to add a placeholder of the jax frontend. It initializes the backbone of it but doesn't guarantee stable functionalities and APIs. In fact, since lots of codes were borrowed from the pytorch frontend, much work is required in the future.

### Components
 - The jax python binding that is included in the wheel.
 - The cpp-level components, including decoder, translate session, framework node, op table, etc.
 - Necessary changes on cmake files. 
 - Support for ovc, which enables the `ov.convert_model` for jax.
 - Some placeholders, such as an `add` operator copied from pytorch frontend.


### TODO (not sure if they're required in this PR)
- Add the compiling & test of jax frontend to the github workflow.
- Remove some duplicated code that are not necessary for jax frontend.
- Add some unit test.


In my opinion, this PR is already extra-large and probably should be separated into several small PRs for easy review. However, I still opened this PR because it provides the ability to run a simplest jax module, which cannot be done by a single smaller PR. When developing this, I held the view that I should make a smallest example work ASAP, and that's why I copied many codes from pytorch frontend. With this initial work, I can clear the unnecessary parts and add functionalities step by step in the future.

I test it with the following code and it worked well.

```py
import openvino as ov
from openvino.frontend.jax import jax_debug

import jax
import jax.numpy as jnp

@jax_debug
def log2(x, y):
  return jnp.add(x, y)


if __name__ == '__main__':
    ov_model = ov.convert_model(log2, example_input=[jnp.ones([1], dtype=jax.numpy.float32), jnp.zeros([1], dtype=jax.numpy.float32)])
    
    core = ov.Core()
    compiled_model = core.compile_model(ov_model, 'CPU')
    output = compiled_model({0: 5.0, 1: 3.0})
    print(output) # output: 8.0
```

Here, `jax_debug` is an attribute that is used temporarily in `ovc` to avoid taking a normal function as jax module (jax often takes function as module!). It will be removed in the future.

@rkazants Could you please help to take a quick look of this PR? I'd like to separate this PR into some smaller ones for easy review before or after our meeting tomorrow. 

---

BTW, may I ask for creating a new label named `category: JAX FE`?

**Ticket:** 142304

rkazants commented:
build_jenkins

rkazants commented:
Hi @AsakusaRinne, for now please check CI logs and fix revealed issues.

Best regards,
Roman

AsakusaRinne commented:
Hi, I've made some changes according to the comments and workflow logs. Could you please help to trigger the workflows again?

rkazants commented:
build_jenkins

mvafin commented:
Please add a layer test for a simple `add` model

mvafin commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,149 @@
+        
+        # TODO: the `self.jaxpr` here is possible to be a jaxpr or a jax_eqn. Maybe we need a better design.
+        if jax_eqn is None:
+            self.jaxpr = jax.make_jaxpr(self.jax_module)(*self.example_inputs).jaxpr
```
Please move the step `jax.make_jaxpr` to before the `convert_model`. And this way you won't need `jax_debug` function

mvafin commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,149 @@
+        else:
+            return "root"
+        
+    def get_schema(self) -> str:
+        # TODO: this is not really the semantic of schema in jaxpr. Need to check again.
+        return str(self.jaxpr)
```
I don't think this is applicable for JAX, this is pytorch specific function

mvafin commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,149 @@
+    def get_subgraph_decoder(self, index: int) -> Decoder:
+        raise NotImplementedError("Haven't figured out when it should be used.")
+    
+    def get_subgraph_size(self) -> int:
+        if isinstance(self.jaxpr, jax.core.JaxprEqn):
+            return 0
+        else:
+            # TODO: check again if there's no subgraph in jaxpr
+            return 1
```
Please check if nested models possible in JAX, if no, this function is not needed

mvafin commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,149 @@
+        # This may not be necessary in jax frontend but needs further check.
+        raise NotImplementedError("Not implemented yet.")
+    
+    def as_string(self):
+        return str(self.jaxpr)
```
This function was introduced in pytorch FE to convert node to string if it is a constant of type string, here I don't think you do that, maybe this function is not needed for JAX

mvafin commented on src/frontends/jax/src/frontend.cpp:
```diff
@@ -0,0 +1,249 @@
+
+    // the following 2 transformations are needed for keypoint detectron2 models to work.
+    // AtenIndexToSelect will be called twice
+    manager.register_pass<ov::pass::ConvertConvertLike>();
+
+    // Mark quantized and f16/bf16 compressed constants to prevent CF for them,
+    // so that not extra memory is used for intermediate decompressed constants.
+    manager.register_pass<ov::pass::MarkDequantizationSubgraph>(
+        element::TypeVector{element::u8, element::i8, element::u4, element::i4});
+    manager.register_pass<ov::pass::MarkCompressedFloatConstants>();
+    manager.register_pass<ov::pass::ConstantFolding>();
+
+    manager.register_pass<ov::pass::ConvertConvertPromoteTypes>();
+    manager.register_pass<ov::pass::PushConstantToSubgraph>();
+    manager.register_pass<ov::pass::UnrollIf>();
+    // TODO: this is only a fake pass and needs to be removed.
+    manager.register_pass<ov::frontend::jax::pass::PlaceholderInBodyReplacer>();
+    manager.register_pass<ov::pass::RemoveMultiSubGraphOpDanglingParamsResults>();
+    manager.register_pass<ov::pass::ReverseShapeAndTypeInfer>();
+    manager.register_pass<ov::pass::ResolveNameCollisions>(true);
+    manager.run_passes(model);
+
+    // Usually if nn.Module.forward is given as a source model for conversion, there is the first Parameter
+    // that represents original `self` argument in forward(self, ...). `self` shouldn't play any role in model
+    // inference if model is completely frozen and all methods are inlined. So we check if it doesn't have any
+    // consumers in the finally converted model and remove this parameter. This parameter should have index 0.
+    if (model->get_parameters().size() > 0) {
+        auto self = model->get_parameters()[0];
+        if (self->output(0).get_target_inputs().empty()) {
+            // There is no consumers: safe to remove
+            OPENVINO_DEBUG << "[ WARNING ] Removing parameter[0] in converted jax model, because it is never used "
+                              "and treated as `self`\n";
+            model->remove_parameter(self);
+        } else {
+            OPENVINO_DEBUG << "[ WARNING ] Couldn't remove parameter[0] in converted jax model\n";
+        }
+    }
```
Remove all the passes for now, they are not required in general and should be added if needed

rkazants reviewed (commented):
Please do the following:
 - remove unneeded code in Python code
 - add one line comment to describe used members of JaxPr
 - remove unneeded code in C++ part (transformation calls in normalize, )

rkazants commented on src/frontends/jax/src/translate_session.cpp:
```diff
@@ -0,0 +1,384 @@
+
+using ReversepropCreatorFunction = std::function<ov::Output<ov::Node>(const Output<Node>&, const Output<Node>&)>;
+
+Output<Node> TranslateSession::get_reverseprop_op(const std::shared_ptr<JaxDecoder>& node,
+                                                  const Output<Node>& direct_op_output,
+                                                  const Output<Node>& value) {
+    std::map<std::string, ReversepropCreatorFunction> backprop_map = {};
+
+    Output<Node> backprop_node;
+    try {
+        auto it = backprop_map.find(node->get_op_type());
+        if (it != backprop_map.end()) {
+            return it->second(direct_op_output, value);
+        }
+
+    } catch (std::exception& e) {
+        OPENVINO_DEBUG << "Exception happened during conversion of backprop op: " << node->get_op_type()
+                       << " with schema: " << node->get_schema() << ": " << e.what();
+    }
+    // Create PtFrameworkNode representing unconverted backprop operation
+    return std::make_shared<JaxFrameworkNode>(node, OutputVector{value}, 1, true);
+}
```
not needed

mvafin commented on src/frontends/jax/src/op/add.cpp:
```diff
@@ -0,0 +1,82 @@
+    return translate_add_common(context, false);
+};
+
+OutputVector translate_add_(const NodeContext& context) {
```
`add_` is pytorch specific, please remove

mvafin commented:
Please remove all unused code copied from pytorch

rkazants commented on src/frontends/jax/src/op/add.cpp:
```diff
@@ -0,0 +1,82 @@
+        // Case when two lists gets concatenated
+        JAX_OP_CONVERSION_CHECK(false, "aten::add is used for concatenation of lists, not possible to convert");
+    }
+    if (inplace) {
+        lhs = context.get_input(0);
+        rhs = context.get_input(1);
+        if (lhs.get_element_type().is_dynamic() || lhs.get_element_type() != rhs.get_element_type())
+            rhs = context.mark_node(std::make_shared<v1::ConvertLike>(rhs, lhs));
+    } else {
+        std::tie(lhs, rhs) = get_inputs_with_promoted_types(context, 0, 1);
+    }
+
+    auto left_is_bool = lhs.get_element_type() == ov::element::boolean ||
+                        (dtype0.is<element::Type>() && dtype0.as<element::Type>() == element::boolean);
+    auto right_is_bool = rhs.get_element_type() == ov::element::boolean ||
+                         (dtype1.is<element::Type>() && dtype1.as<element::Type>() == element::boolean);
+
+    if (left_is_bool && right_is_bool) {
+        // when types are boolean then add means logical Or operation
+        auto logical_or = context.mark_node(std::make_shared<v1::LogicalOr>(lhs, rhs));
+        if (inplace)
+            context.mutate_input(0, logical_or);
+
+        return {logical_or};
+    }
+
+    Output<Node> alpha;
+    if (!context.input_is_none(2)) {
+        alpha = context.get_input(2);
+    } else if (context.has_attribute("alpha")) {
+        alpha = context.get_attribute<Output<Node>>("alpha");
+    }
+    if (alpha.get_node_shared_ptr()) {
+        auto converted_alpha = context.mark_node(std::make_shared<v1::ConvertLike>(alpha, rhs));
+        rhs = context.mark_node(std::make_shared<v1::Multiply>(converted_alpha, rhs));
+    }
+    auto add = context.mark_node(std::make_shared<v1::Add>(lhs, rhs));
+    if (inplace)
+        context.mutate_input(0, add);
+    return {add};
```
please simplify. remove PT specific code

rkazants commented on src/tests/functional/plugin/conformance/subgraphs_dumper/include/utils/cache.hpp:
```diff
@@ -4,11 +4,11 @@
 
 #pragma once
 
```
please revert changes here

rkazants replied:
please revert changes

AsakusaRinne commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,149 @@
+    def get_output_shape(self, index):
+        return PartialShape(self.jaxpr.outvars[index].aval.shape)
+    
+    def decoder_type_name(self) -> str:
+        return "jaxpr"
```
We can also use HLO (the IR of XLA) to parse the jax modules. Though it is not included in the jax frontend project this summer, do you think I should assume there might be multiple decoders for jax in the future?

mvafin replied:
I suggest remove it for now and add back when needed

rkazants commented on tools/mo/automation/package_BOM.txt:
```diff
@@ -835,6 +835,7 @@ openvino/tools/mo/moc_frontend/__init__.py
 openvino/tools/mo/moc_frontend/analysis.py
 openvino/tools/mo/moc_frontend/check_config.py
 openvino/tools/mo/moc_frontend/extractor.py
+openvino/tools/mo/moc_frontend/jax_frontend_utils.py
```
```suggestion
```
mo != ovc, no need to update it:)

rkazants commented on src/bindings/python/src/pyopenvino/frontend/jax/py_module.cpp:
```diff
@@ -0,0 +1,16 @@
+
+PYBIND11_MODULE(py_jax_frontend, m) {
+    regclass_frontend_jax_decoder(m);
+    // regclass_frontend_pytorch_ConversionExtension(m);
+    // regclass_frontend_pytorch_OpExtension(m);
```
```suggestion
```

rkazants commented on src/frontends/jax/src/utils.hpp:
```diff
@@ -0,0 +1,216 @@
+
+Output<Node> reshape_channelwise(const NodeContext& context,
+                                 const Output<Node>& data,
+                                 const Output<Node>& shape_source);
+
+std::tuple<Output<Node>, Output<Node>> get_shape_rank(const NodeContext& context,
+                                                      const Output<Node>& x,
+                                                      bool as_scalar = false,
+                                                      element::Type output_type = element::i32);
+
+Output<Node> reshape_kernel_for_group(const NodeContext& context, const Output<Node>& kernel, int64_t groups);
+
+std::shared_ptr<Node> get_axes_range(const NodeContext& context, int input_id);
+
+std::shared_ptr<Node> get_node_axes_range(const NodeContext& context, const Output<Node>& x);
+
+Output<Node> normalize_axis(const NodeContext& context, const Output<Node>& axis, const Output<Node>& input_node);
+
+std::shared_ptr<Node> numel(const NodeContext& context,
+                            const Output<Node>& x,
+                            element::Type output_type = element::i32);
+
+element::Type convert_dtype(int64_t dtype_value);
+
+Output<Node> apply_dtype(const NodeContext& context, size_t dtype_port, const Output<Node>& input_tensor);
+
+op::PadType convert_pad(const std::string& pt_pad);
+
+Output<Node> concat_list_construct(const Output<Node>& input);
+
+/// \brief Checks if input represents empty list.
+/// \param input Input to check.
+/// \return true if input is empty list, false - if input is non-empty or non-list.
+bool is_empty_list(const Output<Node>& input);
+
+OutputVector make_framework_node_ignore_bodies(const NodeContext& context, const std::string& exception);
+OutputVector make_framework_node(const NodeContext& context, const std::string& exception);
+
+std::shared_ptr<op::util::FrameworkNode> cast_fw_node(std::shared_ptr<Node> node, const std::string& type);
+
```
please remove unneeded routines

AsakusaRinne replied:
The unneeded routines have been removed.

rkazants commented on tools/ovc/unit_tests/ovc/package_BOM.txt:
```diff
@@ -13,7 +13,8 @@ openvino/tools/ovc/main.py
 openvino/tools/ovc/moc_frontend/analysis.py
 openvino/tools/ovc/moc_frontend/check_config.py
-openvino/tools/ovc/moc_frontend/extractor.py
+openvino/tools/ovc/moc_frontend/extractor.py\
```
```suggestion
openvino/tools/ovc/moc_frontend/extractor.py
```

rkazants commented:
build_jenkins

rkazants commented:
@AsakusaRinne, please fix test failures in JAX - TF FE:

                concrete_func = tf_function.get_concrete_function()
        except Exception as e:
>           raise Exception(error_message.format(e))
E           Exception: Could not trace the TF model with the following error: add got incompatible shapes for broadcasting: (2, 3, 4, 5), (4, 2, 5, 3).

/venv/lib/python3.11/site-packages/openvino/frontend/tensorflow/utils.py:150: Exception
=============================== warnings summary ===============================
jax_tests/test_add.py::TestAdd::test_dot_general[ ie_device:CPU - precision:FP16 - input_type:<class 'numpy.float32'> - params:{'lhs_shape': [4], 'rhs_shape': [4]} ]
  /venv/lib/python3.11/site-packages/tensorflow/python/framework/dtypes.py:35: DeprecationWarning: ml_dtypes.float8_e4m3b11 is deprecated. Use ml_dtypes.float8_e4m3b11fnuz
    from tensorflow.tsl.python.lib.core import pywrap_ml_dtypes

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
---- generated xml file: /__w/openvino/openvino/install/tests/TEST-jax.xml -----
=========================== short test summary info ============================
FAILED install/tests/layer_tests/jax_tests/test_add.py::TestAdd::test_dot_general[ ie_device:CPU - precision:FP16 - input_type:<class 'numpy.float32'> - params:{'lhs_shape': [2, 3, 4, 5], 'rhs_shape': [4, 2, 5, 3]} ] - Exception: Could not trace the TF model with the following error: add got incompatible shapes for broadcasting: (2, 3, 4, 5), (4, 2, 5, 3).
FAILED install/tests/layer_tests/jax_tests/test_add.py::TestAdd::test_dot_general[ ie_device:CPU - precision:FP16 - input_type:<class 'numpy.int32'> - params:{'lhs_shape': [2, 3, 4, 5], 'rhs_shape': [4, 2, 5, 3]} ] - Exception: Could not trace the TF model with the following error: add got incompatible shapes for broadcasting: (2, 3, 4, 5), (4, 2, 5, 3).
FAILED install/tests/layer_tests/jax_tests/test_add.py::TestAddWithConstant::test_dot_general[ ie_device:CPU - precision:FP16 - input_type:<class 'numpy.float32'> - params:{'lhs_shape': [2, 3, 4, 5], 'rhs_shape': [4, 2, 5, 3]} ] - Exception: Could not trace the TF model with the following error: add got incompatible shapes for broadcasting: (2, 3, 4, 5), (4, 2, 5, 3).
FAILED install/tests/layer_tests/jax_tests/test_add.py::TestAddWithConstant::test_dot_general[ ie_device:CPU - precision:FP16 - input_type:<class 'numpy.int32'> - params:{'lhs_shape': [2, 3, 4, 5], 'rhs_shape': [4, 2, 5, 3]} ] - Exception: Could not trace the TF model with the following error: add got incompatible shapes for broadcasting: (2, 3, 4, 5), (4, 2, 5, 3).

ilya-lavrenov commented on src/frontends/jax/src/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+
+ov_add_frontend(NAME jax
+                LINKABLE_FRONTEND
+                SHUTDOWN_PROTOBUF
```
as I see, JAX frontend does not use protobuf, hence `SHUTDOWN_PROTOBUF` is not required.

mvafin commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,191 @@
+        else:
+            return id(self.jaxpr.invars[idx])
+    
+    def get_attribute(self, name):
```
Do jax operations have attributes? Maybe remove this until needed

AsakusaRinne replied:
AFAIK they don't, I've removed it. :)

mvafin commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,191 @@
+        return OVAny(None)
+    
+    def get_input_debug_name(self, index) -> str:
+        return "jaxpr_invar_" + str(index)
```
debug name was pytorch-only primitive. I suggest removing it from jax

mvafin commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,191 @@
+    
+    def get_named_input(self, name):
+        # TODO: check again if there's named input in jaxpr
+        raise NotImplementedError("Currently named input is not expected in jax.")
```
Remove for now, until needed by jax frontend

mvafin commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,191 @@
+        raise NotImplementedError("Currently named input is not expected in jax.")
+        
+    def get_output_debug_name(self, index) -> str:
+        return "jaxpr_outvar_" + str(index)
```
Same comment as for input

mvafin commented on src/bindings/python/src/pyopenvino/frontend/jax/decoder.hpp:
Please leave only the functions that exist in python part

mvafin commented on src/frontends/jax/src/utils.cpp:
```diff
@@ -0,0 +1,177 @@
+}
+
+// helper ops
+Output<Node> masked_fill(ov::pass::NodeRegistry& rg,
```
Please remove

mvafin commented on src/frontends/jax/src/utils.cpp:
```diff
@@ -0,0 +1,177 @@
+    }
+}
+
+void copy_runtime_info_and_name(const std::shared_ptr<Node>& from,
```
Don't think you use this now. Remove until needed

mvafin commented on src/frontends/jax/src/utils.cpp:
```diff
@@ -0,0 +1,177 @@
+    return x;
+}
+
+std::tuple<Output<Node>, Output<Node>> get_inputs_with_promoted_types(const NodeContext& context,
```
Is this needed for Jax?

AsakusaRinne replied:
When two inputs with different types are given to an operation, JAX has similar behaviors with PyTorch, that takes the type with higher precision/range as the output type. If I'm not misunderstanding it, I think we need this API for JAX.

mvafin commented on src/frontends/jax/src/op_table.cpp:
```diff
@@ -0,0 +1,28 @@
+
+// Supported ops for Jaxpr
+const std::map<std::string, CreatorFunction> get_supported_ops_jaxpr() {
+    return {{"add", op::translate_1to1_match_2_inputs<opset10::Add>}, {"constant", op::translate_constant}};
```
Use opset 14 or better use the operation directly (without opset)

mvafin commented on src/frontends/jax/src/translate_session.cpp:
```diff
@@ -0,0 +1,364 @@
+    {
+        auto parameters = std::make_shared<ParameterVector>();
+        auto tensor_map = std::make_shared<TensorMap>();  // tensor map of the current context
+        auto mutated_tensors = std::make_shared<std::set<size_t>>();
```
I would recommend removing mutated tensor logic for now. It may be implemented dufferently in JAX

mvafin commented on src/frontends/jax/src/node_context.cpp:
```diff
@@ -0,0 +1,224 @@
+    return ov_node;
+}
+
+void NodeContext::add_tensor_to_context(size_t index, Output<Node> ov_output) const {
+    if (m_tensor_map->count(index)) {
+        OPENVINO_DEBUG << "[ WARNING ] Current context has tensor " << index << ". Assuming mutated output.\n";
+    }
+    m_translate_session->encode_tensor_name(ov_output, index);
+    (*m_tensor_map)[index] = ov_output;
+}
+
+Output<Node> NodeContext::get_tensor_from_model_or_create_input(size_t index) const {
+    if (m_tensor_map->find(index) != m_tensor_map->end()) {
+        return m_tensor_map->at(index);
+    } else {
+        // nested subgraphs case
+        auto parameter = std::make_shared<v0::Parameter>(element::dynamic, PartialShape::dynamic());
+        m_translate_session->encode_tensor_name(parameter->output(0), index);
+        (*m_tensor_map)[index] = parameter;
+        m_external_parameters->push_back(parameter);
+        OPENVINO_DEBUG << "Nested case, created: " << parameter << '\n';
+        return parameter;
+    }
+}
+
+Output<Node> NodeContext::get_input_from_visible_context(size_t index) const {
+    FRONT_END_GENERAL_CHECK(index < get_input_size(), "Index ", index, " is lower then number of inputs.");
+    auto input_tensor = get_input(static_cast<int>(index));
+    auto input_node = input_tensor.get_node_shared_ptr();
+    if (std::dynamic_pointer_cast<v0::Parameter>(input_node)) {
+        // We need to look into external context for inputs that would be feed into
+        // this parameter
+        size_t tensor_idx = m_translate_session->decode_tensor_name(input_node->output(0));
+        if (m_ext_tensor_map.count(tensor_idx)) {
+            input_tensor = m_ext_tensor_map.at(tensor_idx);
+        }
+    }
+    return input_tensor;
+}
```
Since we do not consider nested models now I would recommend removing all logic for nested models

mvafin commented on src/frontends/jax/src/translate_session.cpp:
```diff
@@ -0,0 +1,364 @@
+            auto raw_inputs = node->inputs();
+            for (size_t i = 0; i < raw_inputs.size(); ++i) {
+                auto input = raw_inputs.at(i);
+                if (tensor_map->find(input) == tensor_map->end()) {
+                    // Input refers value in the outer scope, need to create a new
+                    // Parameter in the current scope Linkage to external scope will be
+                    // performed on the level of the parent operation (if or loop)
+                    // TODO: Eliminate duplication with the main code for Parameters
+                    // creation
+                    PartialShape ps = node->get_input_shape(i);
+                    auto type = simplified_type_interpret(node->get_input_type(i));
+                    auto dtype = element::dynamic;
+                    if (type.is<element::Type>()) {
+                        dtype = type.as<element::Type>();
+                    }
+                    auto parameter = std::make_shared<v0::Parameter>(dtype, ps);
+                    (*tensor_map)[input] = parameter;
+                    // set name of parameter to the index of node in the model
+                    encode_tensor_name(parameter->output(0), input);
+                    parameters->push_back(parameter);
+                    inserted_params.push_back(input);
+                }
```
Please remove this, this is also about nested model scope

AsakusaRinne replied:
This part of code seems to be needed in JAX frontend except `inserted_params`, which has been removed.

mvafin commented on src/frontends/jax/src/translate_session.cpp:
```diff
@@ -0,0 +1,364 @@
+
+            for (size_t i = 0; i < fw_outputs.size(); ++i) {
+                size_t fw_tensor_id = node->output(i);
+                if (node->inputs().size() > 0) {
+                    // TODO: do we need to check other inputs, not only 0?
+                    auto in_tensor_id = node->inputs().at(0);
+                    if (m_may_be_alias.count(fw_tensor_id)) {
+                        size_t recorded_in_tensor_id;
+                        std::shared_ptr<JaxDecoder> recorded_node;
+                        std::tie(recorded_in_tensor_id, recorded_node, std::ignore) = m_may_be_alias.at(fw_tensor_id);
+                        FRONT_END_GENERAL_CHECK(recorded_in_tensor_id == in_tensor_id,
+                                                "Operation ",
+                                                context.get_op_type(),
+                                                " creates alias to tensor which was already created before by ",
+                                                recorded_node->get_op_type(),
+                                                ", but from different tensor: ",
+                                                in_tensor_id,
+                                                " vs ",
+                                                recorded_in_tensor_id);
+                    }
+                    m_may_be_alias[fw_tensor_id] = {node->inputs().at(0), node, converted_outputs[i]};
+                    OPENVINO_DEBUG << "Registered alias: " << fw_tensor_id << " of tensor: " << node->inputs().at(0)
+                                   << " of operation: " << context.get_op_type();
+                }
```
Remove this and `m_may_be_alias`

mvafin commented on src/frontends/jax/src/translate_session.cpp:
```diff
@@ -0,0 +1,364 @@
+                OPENVINO_DEBUG << "Mutated tensor with id " << tensor_id << " doesn't exist in inputs, skipping.";
+            }
+        }
+        if (!external_tensor_map.empty()) {
+            // for internal bodies we want to remove all extra inputs that were
+            // created, but not used
+            parameters->erase(std::remove_if(parameters->begin(),
+                                             parameters->end(),
+                                             [&](std::shared_ptr<v0::Parameter> p) {
+                                                 auto tensor_id = decode_tensor_name(p);
+                                                 return p->output(0).get_target_inputs().empty() &&
+                                                        std::find(inserted_params.begin(),
+                                                                  inserted_params.end(),
+                                                                  tensor_id) != inserted_params.end();
+                                             }),
+                              parameters->end());
+        }
```
Since we do not consider nested models we will not create extra inputs

mvafin commented on src/frontends/jax/src/frontend.cpp:
```diff
@@ -0,0 +1,224 @@
+}
+
+void FrontEnd::normalize(const std::shared_ptr<ov::Model>& model) const {
+    // Usually if nn.Module.forward is given as a source model for conversion,
+    // there is the first Parameter that represents original `self` argument in
+    // forward(self, ...). `self` shouldn't play any role in model inference if
+    // model is completely frozen and all methods are inlined. So we check if it
+    // doesn't have any consumers in the finally converted model and remove this
+    // parameter. This parameter should have index 0.
+    if (model->get_parameters().size() > 0) {
+        auto self = model->get_parameters()[0];
+        if (self->output(0).get_target_inputs().empty()) {
+            // There is no consumers: safe to remove
+            OPENVINO_DEBUG << "[ WARNING ] Removing parameter[0] in converted jax "
+                              "model, because it is never used "
+                              "and treated as `self`\n";
+            model->remove_parameter(self);
+        } else {
+            OPENVINO_DEBUG << "[ WARNING ] Couldn't remove parameter[0] in converted "
+                              "jax model\n";
+        }
+    }
```
This is not needed for JAX since first input is not self.

mvafin commented on src/frontends/jax/src/input_model.cpp:
```diff
@@ -0,0 +1,172 @@
+
+std::vector<ov::frontend::Place::Ptr> InputModel::get_inputs() const {
+    if (m_inputs.size() > 0 && m_inputs[0]) {
+        // We need to remove "self" input to not confuse external users
+        const auto& names = m_inputs[0]->get_names();
+        if (std::any_of(names.cbegin(), names.cend(), [](const std::string& n) {
+                return n.find("self") != std::string::npos;
+            }))
+            return std::vector<ov::frontend::Place::Ptr>(m_inputs.begin() + 1, m_inputs.end());
```
This is not needed since self is not in jax inputs

mvafin commented on src/frontends/jax/src/input_model.cpp:
```diff
@@ -0,0 +1,172 @@
+        return p->is_input();
+    });
+    FRONT_END_GENERAL_CHECK(all_inputs, "Only initial inputs are supported by override_all_inputs.");
+    // We need to add back "self" input if it was in initial inputs
+    if (m_inputs.size() > 0 && m_inputs[0]) {
+        // We need to remove "self" input to not confuse external users
+        const auto& names = m_inputs[0]->get_names();
+        if (std::any_of(names.cbegin(), names.cend(), [](const std::string& n) {
+                return n.find("self") != std::string::npos;
+            })) {
+            FRONT_END_GENERAL_CHECK(inputs.size() == m_inputs.size() - 1,
+                                    "Number of inputs provided is incorrect. Graph modification is not supported for "
+                                    "this model. Expected number of inputs: ",
+                                    m_inputs.size() - 1,
+                                    " received ",
+                                    inputs.size());
+            auto self_place = m_inputs[0];
+            // Verify that no same place already in vector
+            auto no_self = std::all_of(inputs.cbegin(), inputs.cend(), [&](const Place::Ptr& p) {
+                return !p->is_equal(self_place);
+            });
+            FRONT_END_GENERAL_CHECK(no_self, "Unexpected input of 'self' was provided to override_all_inputs.");
+            m_inputs = std::vector<ov::frontend::Place::Ptr>{self_place};
+            m_inputs.insert(m_inputs.cend(), inputs.cbegin(), inputs.cend());
+            return;
+        }
+    }
```
Not really needed since self is not in jax inputs

rkazants commented:
Discussed with Ilya packaging and validation approach and agreed on the following:
- build JAX FE by default
- do not include JAX FE binaries, related files into openvino
- put JAX FE binaries, python sources and pybind into openvino_tests archive produced in GHA
- for running JAX FE layer tests, unpack openvino_tests archive and place JAX FE artifacts along with already installed other FEs files in the installation directory

AsakusaRinne commented:
@rkazants I've addressed the most of the comments but have two questions. Could you please help me?

1. Could you please talk about how to use the operation without `opset` namespace? After our meeting I tried to find it from the implementations of pytorch and tensorflow frontend but failed.
2. About the packaging approach: it's a good and reasonable approach but I wonder if it's possible to make it more simple. We can use an environment variable to decide whether to include the jax-related files in `src/bindings/python/wheel/setup.py`. We set it as off by default while enable it in github workflows that produces openvino_tests archive. In this way the jax binaries will not get published in the python package, either.

rkazants commented on src/bindings/python/src/pyopenvino/frontend/jax/decoder.hpp:
```diff
@@ -0,0 +1,84 @@
+
+namespace py = pybind11;
+
+/// Trampoline class to support inheritence from TorchDecoder in Python
```
```suggestion
/// Trampoline class to support inheritence from JaxDecoder in Python
```

rkazants replied:
```suggestion
/// Trampoline class to support inheritence from JaxDecoder in Python
```

rkazants commented on src/bindings/python/src/openvino/frontend/jax/utils.py:
```diff
@@ -0,0 +1,38 @@
+        raise ValueError(f"Constant is expected to be a numpy array or jax array but got {type(arr)}")
+
+jax_to_ov_type_map = {
+    np.dtypes.Float32DType: OVType.f32,
+    np.dtypes.BoolDType: OVType.boolean,
+    jax.dtypes.bfloat16: OVType.bf16, # TODO: check this
+    np.dtypes.Float16DType: OVType.f16,
+    np.dtypes.Float32DType: OVType.f32,
+    np.dtypes.Float64DType: OVType.f64,
+    np.dtypes.UInt8DType: OVType.u8,
+    np.dtypes.Int8DType: OVType.i8,
+    np.dtypes.Int16DType: OVType.i16,
+    np.dtypes.Int32DType: OVType.i32,
+    np.dtypes.Int64DType: OVType.i64,
```
let us better use native JAX types like  jnp.float32. I would not use numpy dep here

rkazants commented on src/frontends/jax/src/utils.cpp:
```diff
@@ -0,0 +1,544 @@
+    return context.get_input_type(index).is<type::PyScalar>();
+}
+
+void align_eltwise_input_types(const NodeContext& context,
```
I think we should take a look at this https://jax.readthedocs.io/en/latest/type_promotion.html

rkazants replied:
please remove it since it not used

rkazants commented on src/frontends/jax/src/utils.hpp:
```diff
@@ -0,0 +1,149 @@
+namespace frontend {
+namespace jax {
+
+const std::string jax_prefix = "[Jax Frontend] ";
```
```suggestion
const std::string jax_prefix = "[JAX Frontend] ";
```

rkazants replied:
```suggestion
const std::string jax_prefix = "[JAX Frontend] ";
```

rkazants commented on tools/ovc/openvino/tools/ovc/moc_frontend/jax_frontend_utils.py:
```diff
@@ -0,0 +1,26 @@
+def get_jax_decoder(model, args):
+    try:
+        from openvino.frontend.jax.jaxpr_decoder import JaxprPythonDecoder
+        import jax
```
looks not needed import?

rkazants commented on tools/ovc/openvino/tools/ovc/moc_frontend/jax_frontend_utils.py:
```diff
@@ -0,0 +1,26 @@
+# SPDX-License-Identifier: Apache-2.0
+
+import logging as log
+import sys
```
please remove unused imports

rkazants commented on src/bindings/python/src/openvino/frontend/jax/utils.py:
```diff
@@ -0,0 +1,38 @@
+        raise ValueError(f"Constant is expected to be a numpy array or jax array but got {type(arr)}")
+
+jax_to_ov_type_map = {
+    np.dtypes.Float32DType: OVType.f32,
+    np.dtypes.BoolDType: OVType.boolean,
+    jax.dtypes.bfloat16: OVType.bf16, # TODO: check this
+    np.dtypes.Float16DType: OVType.f16,
+    np.dtypes.Float32DType: OVType.f32,
+    np.dtypes.Float64DType: OVType.f64,
+    np.dtypes.UInt8DType: OVType.u8,
+    np.dtypes.Int8DType: OVType.i8,
+    np.dtypes.Int16DType: OVType.i16,
+    np.dtypes.Int32DType: OVType.i32,
+    np.dtypes.Int64DType: OVType.i64,
```
please use native JAX types instead of numpy

AsakusaRinne replied:
AFAIK JAX uses numpy dtype directly except of bfloat16. :)

rkazants replied:
agreed to adjust it since it exists for JAX

rkazants commented on tools/ovc/openvino/tools/ovc/moc_frontend/jax_frontend_utils.py:
```diff
@@ -0,0 +1,26 @@
+        from openvino.frontend.jax.jaxpr_decoder import JaxprPythonDecoder
+        import jax
+    except Exception as e:
+        log.error("Jax frontend loading failed")
```
```suggestion
        log.error("JAX frontend loading failed")
```

rkazants commented on tools/ovc/openvino/tools/ovc/moc_frontend/jax_frontend_utils.py:
```diff
@@ -0,0 +1,26 @@
+# Copyright (C) 2018-2024 Intel Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+import logging as log
+import sys
+
+import numpy as np
+# pylint: disable=no-name-in-module,import-error
+from openvino.runtime import Tensor, PartialShape
+from openvino.tools.ovc.error import Error
```
please remove unused imports. Also, check other py files to this problem

mvafin commented on tests/layer_tests/jax_tests/jax_layer_test_class.py:
```diff
@@ -11,14 +12,25 @@
+    
+    @staticmethod
+    def use_jaxpr_tracing():
+        torch_compile_env = os.getenv("JAX_TRACE_MODE")
+        if torch_compile_env is not None:
+            return torch_compile_env == "JAXPR"
```
```suggestion
        jax_compile_env = os.getenv("JAX_TRACE_MODE")
        if jax_compile_env is not None:
            return jax_compile_env == "JAXPR"
```

rkazants commented on .github/workflows/job_python_unit_tests.yml:
```diff
@@ -246,6 +246,14 @@ jobs:
+      
+      - name: JAX Layer Tests - JAX FE
+        if: ${{ fromJSON(inputs.affected-components).JAX_FE.test && runner.arch != 'ARM64' && runner.os != 'macOS' }}
+        run: python3 -m pytest ${LAYER_TESTS_INSTALL_DIR}/jax_tests/ -m precommit_jax_fe --junitxml=${INSTALL_TEST_DIR}/TEST-jax.xml
```
```suggestion
        run: python3 -m pytest ${LAYER_TESTS_INSTALL_DIR}/jax_tests/ -m precommit_jax_fe --junitxml=${INSTALL_TEST_DIR}/TEST-jax_fe.xml
```

rkazants replied:
the same file is produced for TF FE

mvafin commented on tests/layer_tests/jax_tests/test_add.py:
Please add tests for mixed types

mvafin commented on src/frontends/jax/src/utils.hpp:
```diff
@@ -0,0 +1,149 @@
+template <typename T>
+OutputVector translate_1to1_match_2_inputs_align_types(const NodeContext& context) {
+    num_inputs_check(context, 2, 2);
+    auto lhs = context.get_input(0);
+    auto rhs = context.get_input(1);
+    auto lhs_type = context.get_input_type(0);
+    auto rhs_type = context.get_input_type(1);
+    // If type is string or None, we shouldn't align
+    if (!lhs_type.is<type::Str>() && !rhs_type.is<type::Str>() && !lhs_type.is<type::PyNone>() &&
+        !rhs_type.is<type::PyNone>()) {
+        align_eltwise_input_types(context,
+                                  lhs,
+                                  rhs,
+                                  is_python_scalar_input(context, 0),
+                                  is_python_scalar_input(context, 1));
+    }
+    OutputVector res = {context.mark_node(std::make_shared<T>(lhs, rhs))};
+    align_output_types(context, res);
+    return res;
+}
+
+template <typename T, size_t idx = 0>
+OutputVector inplace_translate_1to1_match_2_inputs_align_types(const NodeContext& context) {
+    num_inputs_check(context, 2, 2);
+    auto lhs = context.get_input(0);
+    auto rhs = context.get_input(1);
+    // For inplace op we know direction of type alignment
+    if (lhs.get_element_type().is_dynamic() || lhs.get_element_type() != rhs.get_element_type())
+        rhs = context.mark_node(std::make_shared<ov::op::v1::ConvertLike>(rhs, lhs));
+    OutputVector res = {context.mark_node(std::make_shared<T>(lhs, rhs))};
+    return res;
+}
+
+inline OutputVector return_false_scalar(const NodeContext& context) {
+    return {context.mark_node(ov::op::v0::Constant::create(element::boolean, Shape{}, {false}))};
+}
+
+inline OutputVector skip_node(const NodeContext& context) {
+    return {context.get_input(0)};
+}
```
Remove unused helpers for now

mvafin commented on src/frontends/jax/src/utils.hpp:
```diff
@@ -0,0 +1,149 @@
+                               Output<Node>& rhs,
+                               const bool& is_lhs_python_scalar = false,
+                               const bool& ir_rhs_python_scalar = false);
+void align_output_types(const NodeContext& context, OutputVector& outputs);
+
+Output<Node> get_input_with_floating_type(const NodeContext& context, size_t idx);
+
+Output<Node> get_input_as_i32(const NodeContext& context, size_t idx);
```
These are also unused for now, please remove

mvafin commented:
Please update PR description with new code to use jax frontend.

ilya-lavrenov commented:
build_jenkins

mvafin commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,288 @@
+        self.params = {}
+        if hasattr(self.jaxpr, 'params') and isinstance(self.jaxpr.params, dict):
+            for k in self.jaxpr.params.keys():
+                self.params[k] = self.convert_param_to_constant_node(self.jaxpr, k)
```
named param seems to be similar to named input in FX decoder. I don't completely like how it was done in FX decoder, we had to align API with TS in some way, maybe you could do better since you don't have other decoders to align with? Can it be non-constant? For example output of another operation?

AsakusaRinne replied:
I have checked nearly all the functions in `jax.lax` and didn't find such case, though I'm not 100% sure that this case does not exist because in the source code of JAX, the type of `params` in `JaxEqn` is marked as `Any`. May I treat it as constant in this PR and deal with it if meeting non-constant case in the future?

mvafin replied:
leave it for now as is, but if there is really always a constant you could use the `get_attribute` API that is also used in TF and other FEs.

mvafin commented on src/frontends/jax/src/translate_session.cpp:
```diff
@@ -0,0 +1,265 @@
+            results.push_back(result);
+        }
+
+        // Since parameters can be added we need to list all current parameters
+        std::set<size_t> param_names;
+        for (const auto& param : *parameters) {
+            auto input_idx = decode_tensor_name(param->output(0));
+            param_names.insert(input_idx);
+        }
```
This code is not needed

rkazants commented on cmake/features.cmake:
```diff
@@ -137,6 +137,7 @@ endif()
 ov_option(ENABLE_OV_PADDLE_FRONTEND "Enable PaddlePaddle FrontEnd" ON)
 ov_option(ENABLE_OV_IR_FRONTEND "Enable IR FrontEnd" ON)
 ov_option(ENABLE_OV_PYTORCH_FRONTEND "Enable PyTorch FrontEnd" ON)
+ov_option(ENABLE_OV_JAX_FRONTEND "Enable Jax FrontEnd" ON)
```
```suggestion
ov_option(ENABLE_OV_JAX_FRONTEND "Enable JAX FrontEnd" ON)
```

rkazants commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,310 @@
+        elif isinstance(self.jaxpr, jax.core.JaxprEqn):
+            idx = 0
+            res = []
+            for inp in self.jaxpr.invars:
```
in this case we don't need separate `idx` var

rkazants commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,310 @@
+        if isinstance(self.jaxpr, jax.core.Var):
+            return []
+        elif isinstance(self.jaxpr, jax.core.JaxprEqn):
+            idx = 0
```
```suggestion
```

rkazants commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,310 @@
+            for inp in self.jaxpr.invars:
+                if isinstance(inp, jax.core.Literal):
+                    res.append(self.literals[idx].output(0))
+                    idx += 1
```
```suggestion
```

rkazants commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,310 @@
+        if isinstance(self.jaxpr, jax.core.Var):
+            raise IndexError("The jaxpr is a constant, which does not have input.")
+        else:
+            return id(self.jaxpr.invars[idx])
```
should we handle a case of `Literal`?

rkazants commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,310 @@
+        if isinstance(self.jaxpr, jax.core.Var):
+            raise IndexError("The jaxpr is a constant, which does not have input shape.")
+        else:
+            return PartialShape(self.jaxpr.invars[index].aval.shape)
```
should we handle a case of Literal?

rkazants commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,310 @@
+        else:
+            raise ValueError("This is not a constant node so it cannot be converted to a constant.")
+
+    def input_is_none(self, index) -> bool:
+        return self.jaxpr.invars[index] is None
+        
```
```suggestion
```

rkazants commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,310 @@
+    def as_constant(self):
+        return self.constant
+
+    def input_is_none(self, index) -> bool:
+        return self.constant is None
```
```suggestion
```

rkazants commented on src/bindings/python/src/openvino/frontend/jax/utils.py:
```diff
@@ -0,0 +1,124 @@
+    np.dtypes.Float32DType: OVType.f32,
+    np.dtypes.Float64DType: OVType.f64,
+    np.dtypes.UInt8DType: OVType.u8,
+    np.dtypes.Int8DType: OVType.i8,
+    np.dtypes.Int16DType: OVType.i16,
+    np.dtypes.Int32DType: OVType.i32,
+    np.dtypes.Int64DType: OVType.i64,
+}
+
+jax_to_ov_type_map = {
+    jnp.float32: OVType.f32,
+    jnp.bool: OVType.boolean,
+    jnp.bfloat16: OVType.bf16, # TODO: check this
+    jnp.float16: OVType.f16,
+    jnp.float64: OVType.f64,
+    jnp.uint8: OVType.u8,
+    jnp.int8: OVType.i8,
+    jnp.int16: OVType.i16,
+    jnp.int32: OVType.i32,
+    jnp.int64: OVType.i64,
+}
+
+basic_to_ov_type_map = {
+    int: OVType.i64,
+    float: OVType.f32,
+    bool: OVType.boolean,
+}
+
+ov_type_to_int_map = {
+    OVType.u8: 0,
+    OVType.i8: 1,
+    OVType.i16: 2,
+    OVType.i32: 3,
+    OVType.i64: 4,
+    OVType.f16: 5,
+    OVType.f32: 6,
+    OVType.f64: 7,
+    OVType.boolean: 11,
+    OVType.bf16: 15,
+}
```
please move these dicts to top of this file. It is just for readability

rkazants commented on src/bindings/python/src/openvino/frontend/jax/utils.py:
```diff
@@ -0,0 +1,124 @@
+            op.Constant(OVType.i64, Shape([]), [ov_type_to_int_map[ov_dtype_value]]).outputs()
+    
+    print(f"[WARNING][JAX FE] Cannot get constant from value {ivalue}")
+    return OVAny(None), None, None
```
please return just constant node here because you can get its type and shape easily.

rkazants commented on src/bindings/python/src/pyopenvino/frontend/jax/decoder.hpp:
```diff
@@ -0,0 +1,105 @@
+        PYBIND11_OVERRIDE_PURE(std::vector<size_t>&, JaxDecoder, outputs);
+    }
+
+    std::shared_ptr<ov::Node> mark_node(std::shared_ptr<ov::Node> ov_node) const override {
+        PYBIND11_OVERRIDE_PURE(std::shared_ptr<ov::Node>, JaxDecoder, mark_node, ov_node);
+    }
```
please remove it if it not needed for production

rkazants commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,310 @@
+        else:
+            return get_ov_type_for_value(self.jaxpr.invars[index])
+        
+    def get_named_param(self, name):
```
please give description for JAX FE methods (that are not a part of common FE API)

rkazants commented on src/frontends/jax/include/openvino/frontend/jax/decoder.hpp:
```diff
@@ -0,0 +1,80 @@
+
+    virtual OutputVector get_named_param_as_constant(const std::string& name) const = 0;
+
+    virtual PartialShape get_named_param_shape(const std::string& name) const = 0;
+
+    virtual Any get_named_param_type(const std::string& name) const = 0;
+
```
please remove it if possible

rkazants commented on src/frontends/jax/src/input_model.hpp:
```diff
@@ -0,0 +1,51 @@
+class Place;
+class JaxDecoder;
+
+struct PlaceDesc {
+    PlaceDesc(const std::shared_ptr<Node>& value) : m_value(value) {}
+    std::shared_ptr<Node> m_value;
+};
```
please remove it. Looks unuseful

rkazants commented on src/frontends/jax/src/input_model.hpp:
```diff
@@ -0,0 +1,51 @@
+    std::unordered_map<std::string, std::shared_ptr<frontend::Place>> m_name_to_place;
+    std::vector<std::shared_ptr<frontend::Place>> m_inputs;
+    std::vector<std::shared_ptr<frontend::Place>> m_outputs;
+    std::unordered_map<size_t, PlaceDesc> m_descriptors;
```
please remove it

rkazants commented on src/frontends/jax/src/jax_framework_node.hpp:
```diff
@@ -0,0 +1,79 @@
+class JaxFrameworkNode : public ov::op::util::FrameworkNode {
+public:
+    OPENVINO_OP("JaxFrameworkNode", "util", ::ov::op::util::FrameworkNode);
+    static constexpr const char* op_type_key = "PtTypeName";
+    static constexpr const char* failed_conversion_key = "PtException";
```
```suggestion
    static constexpr const char* op_type_key = "JaxTypeName";
    static constexpr const char* failed_conversion_key = "JaxException";
```

rkazants commented on tools/ovc/openvino/tools/ovc/convert_impl.py:
```diff
@@ -462,6 +471,12 @@ def _convert(cli_parser: argparse.ArgumentParser, args, python_api_used):
+                if get_jax_decoder is not None:
+                    get_jax_decoder(args['input_model'], args)
+                else:
+                    raise Error("JAX decoder is not available.")
```
```suggestion
                    raise Error("JAX Frontend is not available.")
```

rkazants commented on tests/layer_tests/jax_tests/test_add.py:
```diff
@@ -0,0 +1,124 @@
+
+class TestAdd(JaxLayerTest):
+    def _prepare_input(self):
+        lhs = jnp.array(np.random.randint(-10, 10, self.lhs_shape).astype(self.lhs_type))
+        rhs = jnp.array(np.random.randint(-10, 10, self.rhs_shape).astype(self.rhs_type))
```
let us differentiate data generation for floating-point type and integer type.
For floating-point type, we want to have data values like 1.5 and 2.1

rkazants commented on src/frontends/jax/src/utils.cpp:
```diff
@@ -0,0 +1,133 @@
+};
+}  // namespace
+
+element::Type convert_dtype(int64_t pt_type) {
+    FRONT_END_OP_CONVERSION_CHECK(JAX_TO_OV_TYPE.count(pt_type), "Unknown type: ", pt_type);
+    return JAX_TO_OV_TYPE.at(pt_type);
+};
```
```suggestion
element::Type convert_dtype(int64_t jax_type) {
    FRONT_END_OP_CONVERSION_CHECK(JAX_TO_OV_TYPE.count(jax_type), "Unknown type: ", jax_type);
    return JAX_TO_OV_TYPE.at(jax_type);
};
```

rkazants commented on src/frontends/jax/src/utils.cpp:
```diff
@@ -0,0 +1,133 @@
+
+namespace {
+const std::unordered_map<int64_t, element::Type> JAX_TO_OV_TYPE{
+    {0, element::u8},
+    {1, element::i8},
+    {2, element::i16},
+    {3, element::i32},
+    {4, element::i64},
+    {5, element::f16},
+    {6, element::f32},
+    {7, element::f64},
+    {11, element::boolean},
+    {12, element::i8},   // quantized i8
+    {13, element::u8},   // quantized u8
+    {14, element::i32},  // quantized i32
+    {15, element::bf16},
+};
```
let us have this dicr in one place in Python and avoid duplication

rkazants commented on src/frontends/jax/src/utils.cpp:
```diff
@@ -0,0 +1,133 @@
+    return jax_prefix;
+}
+
+bool is_none_node(const Output<Node>& node) {
+    if (const auto& fw_node_inp = std::dynamic_pointer_cast<ov::op::util::FrameworkNode>(node.get_node_shared_ptr())) {
+        const auto& attrs = fw_node_inp->get_attrs();
+        if (attrs.find("none_value") != attrs.end()) {
+            return true;
+        }
+    }
+    return false;
+}
```
please remove it if not used

rkazants commented on src/frontends/jax/src/utils.cpp:
```diff
@@ -0,0 +1,133 @@
+    return false;
+}
+
+Any simplified_type_interpret(Any type) {
+    // Type in jaxpr is already the dtype.
+    return type;
+}
```
looks useless please remove

ilya-lavrenov commented:
Please, extend the following files:
- Labeler with patterns for JAX FE https://github.com/openvinotoolkit/openvino/blob/master/.github/labeler.yml#L141-L162
- Smart CI components similarly to https://github.com/openvinotoolkit/openvino/blob/master/.github/components.yml#L129-L134. I suppose `revalidate MO` is not required, since JAX FE is not part of MO, right? 
- Codeowner files https://github.com/openvinotoolkit/openvino/blob/master/.github/CODEOWNERS#L83-L90 with https://github.com/orgs/openvinotoolkit/teams/openvino-jax-frontend-maintainers group

rkazants commented:
> Please, extend the following files:
> 
> * Labeler with patterns for JAX FE https://github.com/openvinotoolkit/openvino/blob/master/.github/labeler.yml#L141-L162
> * Smart CI components similarly to https://github.com/openvinotoolkit/openvino/blob/master/.github/components.yml#L129-L134. I suppose `revalidate MO` is not required, since JAX FE is not part of MO, right?
> * Codeowner files https://github.com/openvinotoolkit/openvino/blob/master/.github/CODEOWNERS#L83-L90 with https://github.com/orgs/openvinotoolkit/teams/openvino-jax-frontend-maintainers group

Let us do this in separate PR by us.

Best regards,
Roman

rkazants commented:
Hi @AsakusaRinne,

Please apply my latest comments and we are ready to merge PR asap.
We agreed do not exclude JAX FE from openvino package in master and exclude it only in the upcoming release branches. So no additional changes related to packaging are required.

Best regards,
Roman

rkazants commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,310 @@
+        else:
+            return "root"
+        
+    def mark_node(self, node):
+        name = self.get_op_type()
+        if "FrameworkNode" not in node.get_type_name():
+            name += "/" + node.get_type_name()
+        node.set_friendly_name(self.name + "/" + name)
+        return node
```
please leave it since it presents in abstract class

AsakusaRinne commented:
@rkazants Hi, Roman, I've addressed all the comments except one. The `element::Type` cannot be used as the key of `std::unordered_map` because the absence of its hash function. I added a comment in the c++ code to reminder that the python code should be changed when the c++ map is changed. Do you think there's a better way?

Besides, in `jaxpr_decoder.py`, I made some changes to make the design more clear. The constvar, literal and named parameters are all presented by `_JaxprPythonConstantDecoder` now.


```
/usr/include/c++/11/bits/hashtable_policy.h:1217:23: error: static assertion failed: hash function must be invocable with an argument of key type
 1217 |         static_assert(__is_invocable<const _Hash&, const _Key&>{},
```

rkazants commented:
build_jenkins

rkazants commented:
build_jenkins

AsakusaRinne commented:
build_jenkins

AsakusaRinne commented:
@rkazants There was an error possibly caused by JAX version and I added a fix for it. Could you please help to trigger the `build_jenkins` again?

mlukasze commented:
build_jenkins

mvafin commented on src/bindings/python/src/openvino/frontend/jax/jaxpr_decoder.py:
```diff
@@ -0,0 +1,283 @@
+            raise TypeError( f"The input should be a literal or jax array, but got {type(literal)}.")
+        return _JaxprPythonConstantDecoder(constant=constant, name=name, output_id=output_id)
+        
+class _JaxprPythonConstantDecoder (Decoder):
```
Do we really need a separate decoder for constant? We have `decoder.as_constant()` which can work as constant decoder

rkazants replied:
Agree with Maxim, looks not needed.
@AsakusaRinne, please fix it in separate PR.

rkazants commented:
build_jenkins
